lib/fs.js

```diff
@@ -453,42 +454,103 @@ function closeSync(fd) {
-    flags = 'r';
+    callback = flag;
+    flag = 'r';
     mode = 0o666;
   } else if (typeof mode === 'function') {
     callback = mode;
     mode = 0o666;
-  } else {
-    mode = parseFileMode(mode, 'mode', 0o666);
   }
-  const flagsNumber = stringToFlags(flags);
+  // Handle case where 2nd argument is options-bag
+  if (typeof flag === 'object' && flag) {
+    options = flag;
+    flag = options.flag ?? 'r';
+    mode = options.mode ?? 0o666;
+    parents = options.parents;
+  }
+  flag = stringToFlags(flag);
+  mode = parseFileMode(mode, 'mode', 0o666);
   callback = makeCallback(callback);
 
   const req = new FSReqCallback();
-  req.oncomplete = callback;
+  req.oncomplete = (err, fd) => {
+    // Lazily create parent-subdirectories if they do not exist
+    if (err?.code === 'ENOENT' && parents && (flag & O_CREAT)) {
+      mkdir(pathModule.dirname(path), { recursive: true }, (err2) => {
+        // Ignore EEXIST race-condition from multiple, async mkdir()
+        if (err2 && err2.code !== 'EEXIST') {
+          callback(err2);
+          return;
+        }
+        // Retry open() after lazily creating parent-subdirectories
+        open(path, {
+          flag,
+          mode
+        }, callback);
+      });
+      return;
+    }
+    callback(err, fd);
+  };
 
   binding.open(pathModule.toNamespacedPath(path),
-               flagsNumber,
+               flag,
                mode,
                req);
 }
 
 
-function openSync(path, flags, mode) {
+// This function has two possible signatures:
+// fs.openSync(path[, flag[, mode]])
+// fs.openSync(path[, options])
+function openSync(path, flag, mode) {
+  let options;
+  let parents;
   path = getValidatedPath(path);
-  const flagsNumber = stringToFlags(flags);
+  // Handle case where 2nd argument is options-bag
+  if (typeof flag === 'object' && flag) {
+    options = flag;
+    flag = options.flag ?? 'r';
+    mode = options.mode ?? 0o666;
+    parents = options.parents;
+  }
+  flag = stringToFlags(flag);
   mode = parseFileMode(mode, 'mode', 0o666);
 
   const ctx = { path };
   const result = binding.open(pathModule.toNamespacedPath(path),
-                              flagsNumber, mode,
+                              flag, mode,
                               undefined, ctx);
-  handleErrorFromBinding(ctx);
-  return result;
+  try {
+    handleErrorFromBinding(ctx);
+    return result;
+  } catch (err) {
+    // Lazily create parent-subdirectories if they do not exist
+    if (err?.code === 'ENOENT' && parents && (flag & O_CREAT)) {
+      try {
+        mkdirSync(pathModule.dirname(path), { recursive: true });
+      } catch (err2) {
+        // Ignore EEXIST race-condition from multiple, async mkdir()
+        if (err2.code !== 'EEXIST') {
+          throw err2;
+        }
+      }
+      // Retry openSync() after lazily creating parent-subdirectories
+      return openSync(path, {
+        flag,
+        mode
+      });
+    }
+    throw err;
+  }
```

Comment:
We could use `hasNoEntryError` instead of a more expensive `try`/`catch`. We may want to also create a `hasNoEEXIST` function to get rid of the second `try`/`catch`.
```suggestion
  if (parents && (flag & O_CREAT) && hasNoEntryError(ctx)) {
    try {
        mkdirSync(pathModule.dirname(path), { recursive: true });
      } catch (err2) {
        // Ignore EEXIST race-condition from multiple, async mkdir()
        if (err2.code !== 'EEXIST') {
          throw err2;
        }
      }
      // Retry openSync() after lazily creating parent-subdirectories
      return openSync(path, {
        flag,
        mode
      });
    }

  handleErrorFromBinding(ctx);
  return result;
```

Reply:
k, applied changes:

1. replaced 1st try...catch with `hasNoEntryError` check

2. removed 2nd try...catch (and equivalent in `fs.open` and `fs.promises.open`), since to my knowledge `mkdir[Sync](...,{recursive:true})` no longer raises `EEXIST`.  or am i wrong on that?